package/network/ipv6/map/files/map.sh

```diff
@@ -140,19 +141,61 @@ proto_map_setup() {
 	      json_add_string snat_ip $(eval "echo \$RULE_${k}_IPV4ADDR")
 	    json_close_object
 	  else
-	    for portset in $(eval "echo \$RULE_${k}_PORTSETS"); do
-              for proto in icmp tcp udp; do
-	        json_add_object ""
-	          json_add_string type nat
-	          json_add_string target SNAT
-	          json_add_string family inet
-	          json_add_string proto "$proto"
-                  json_add_boolean connlimit_ports 1
-                  json_add_string snat_ip $(eval "echo \$RULE_${k}_IPV4ADDR")
-                  json_add_string snat_port "$portset"
-	        json_close_object
-              done
-	    done
+	    if [ "$snat_fix" = "1" ]; then
+			# SNAT Fix: Get all ports and full port count, excluding DONT_SNAT_TO ports
+			DONT_SNAT_TO="${dont_snat_to:-0}"
+			local portcount=0
+			local allports=""
+			for portset in $(eval "echo \$RULE_${k}_PORTSETS"); do
+				local startport=$(echo $portset | cut -d'-' -f1)
+				local endport=$(echo $portset | cut -d'-' -f2)
+				for x in $(seq $startport $endport); do
+					if ! echo "$DONT_SNAT_TO" | tr ' ' '\n' | grep -qw $x; then
```

Comment:
This part of the loop will be ... slow. Recommend optimising DONT_SNAT_TO to an array. You're making the new variable anyway, so you have the option to use an array in uci. This avoids grep tr and echo calls on every loop iteration.